---
title: "Class 13 Transcriptomics and the analysis of RNA-Seq data"
author: "Eric Wang A17678188"
format: pdf
toc: TRUE
---

## Background

Today we will perform an RNASeq analysis on the effects of dexamethasone (hereafter "dex"), a common steroid, on airway smooth muscle (ASM) cell lines. 

## Data Import

We need two things for this analysis: 

- **countData**: a table with genes as rows and samples/experiments as columns, 
- **colData**: metadata about the columns (i.e. samples) in the main countData object. 

```{r import}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <- read.csv("airway_metadata.csv")
```

Let's have a wee peak at these two objects:
```{r}
metadata
```

```{r}
head(counts)
```

## Check on metadata counts correspondance

We need to check that the metadata matches the samples in our count data. 

```{r}
ncol(counts) == nrow(metadata)
colnames(counts) == metadata$id
```

```{r}
all( c(T,T,T,T))
```

>Q1. how many genes are in this dataset?

```{r}
nrow(counts)
```

>Q2. How many "control" samples are in this dataset? 

```{r}
sum(metadata$dex == "control")
```

## Analysis Plan... 

We have 4 replicates per condition ("control" and "treated"). We want to compare the control vs the treated to see which genes expression levels change when we have the drug present.

We will go row by row (gene by gene) and see if the average value in control columns is different than the average value in treated columns

- Step 1. Find which columns in `counts` correspond to "control" samples. 

- Step 2. Extract/select these columns

- Step 3. Calculate an average value for each gene (i.e. each row).

```{r}
# The indices (i.e positions) that are "control" 
control.inds <- metadata$dex == "control"
```

```{r}
# Extract/select these "control" columns from counts
control.counts <- counts[,control.inds]
```

```{r}
# Calculate the mean for each gene (i.e row)
control.mean <- rowMeans(control.counts)
```

> Q. Do the same for "treated" samples - find the mean count value per gene

```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[,treated.inds]
treated.mean <- rowMeans(treated.counts)
```

Let's put these two mean values into a new data.frame `meancounts` for easy book-keeping and plotting. 

```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)
```

> Q. Make a ggplot of average counts of control vs treated. 

```{r}
library(ggplot2)
ggplot(meancounts, aes(control.mean, treated.mean)) + geom_point(alpha = 0.3) + scale_x_log10() + scale_y_log10()
```



## Log2 units and fold change

If we consider "treated"/ "control" counts we will get a number that tells us the change. 

```{r}
# No change
log2(20/20)
```

```{r}
# A doubling in the treated vs control
log2(40/20)
```

```{r}
log2(10/40)
```

> Q. Add a new column `log2fc` for log2 fold change of treated/control to our `meancounts` object. 

```{r}
meancounts$log2fc <- 
  log2(meancounts$treated.mean/meancounts$control.mean)

head(meancounts)
```

## Remove zero count genes

Typically we would not consider zero count genes - as we have no data about them and they should be excluded from further consideration. These lead to "funky" log2 fold change values (e.g. divide by zero errors etc.)


## DESeq analysis 

We are missing any measure of significance from the work we had so far. Let's do this properly with the **DESeq2** package. 


```{r, message = FALSE}
library(DESeq2)
```


The DESeq2 package, like many bioconductor packages, wants it's input in a very specific way - a data structure setup with all the info it needs for the calculation. 

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, colData = metadata, design = ~dex)
```

The main function in this package is called `DESeq()` it will run the full analysis for us on our `dds` input object: 


```{r}
dds <- DESeq(dds)
```


Extract our results: 

```{r}
res <-results(dds)
head(res)
```


## Volcano plot

A useful summary figure of our results is often called a volcano pot. It is basically a plot of log2 fold change values vs Adjusted p-values. 


> Q. use ggplot to make a first version "volcano plot" of `log2FoldChange` vs `padj` 

```{r}
ggplot(res, aes(log2FoldChange, padj)) + geom_point()
```

This is not very useful because the y-axis (p-value) is not really helpful - we want to focus on low p-values

```{r}
ggplot(res, aes(log2FoldChange, log(padj))) + geom_point()
```


```{r}
ggplot(res, aes(log2FoldChange, -log(padj))) + geom_point() + geom_vline(xintercept = c(-2,+2), col = "red") + geom_hline(yintercept = -log(0.05), col = "red")
```


## Add some plot annotation 

> Q. Add color to the points (genes) we care about, nice axis labels, a useful title and a nice theme. 

```{r}
mycols <- rep("lightblue", nrow(res))
mycols[res$log2FoldChange > 2] <- "yellow"
mycols[res$log2FoldChange < -2] <- "lightgreen"
mycols[res$padj >= 0.05] <- "lightblue"
ggplot(res, aes(log2FoldChange, -log10(padj))) +
  geom_point(col = mycols) +
  geom_vline(xintercept = c(-2, 2), color = "red") +
  geom_hline(yintercept = -log10(0.05), color = "red") +
  labs(
    x = "log2 Fold Change",
    y = "-log10 adjusted p-value",
    title = "Volcano Plot of Differential Expression"
  ) +
  theme_minimal()
```

## Save our results to a CSV file

```{r}
write.csv(res,file = "results.csv")
```


## Add annotation Data

To make sense of our results we need to know what the differentially expressed genes are and what biological pathways and process they are involved in. 



```{r}
head(res)
```

Let's start by mapping our ENSEMBLE ids to the more conventional gene SYMBOL. 

We will use two bioconductor packages for this "mapping" **AnnotationDbi** and **org.Hs.eg.db**. 

We will first need to install these from bioconductor with `BiocManager::install("")` 

```{r}
library(AnnotationDbi)
library(org.Hs.eg.db)
```
```{r}
columns(org.Hs.eg.db)
```


```{r}
res$symbol <- mapIds(org.Hs.eg.db,
       keys = rownames(res), # Our ids
       keytype = "ENSEMBL", # Their format
       column= "SYMBOL") # What I want to translate to 
```

```{r}
head(res)
```

> Q. Can you add "GENENAME" and "ENTREZID as new columns to `res` nmaed "name" and "entrez"?

```{r}
res$name <- mapIds(org.Hs.eg.db,
       keys = rownames(res), # Our ids
       keytype = "ENSEMBL", # Their format
       column= "GENENAME") # What I want to translate to 
res$entrez <- mapIds(org.Hs.eg.db,
       keys = rownames(res), # Our ids
       keytype = "ENSEMBL", # Their format
       column= "ENTREZID") # What I want to translate to 
```

```{r}
head(res)
```



```{r}
write.csv(res, file="results_annotated.csv")
```


## Pathway analysis 

Now we know the gene names (gene symbols) and their entrez IDs we can find out what pathways they are involved in. This is called "pathway analysis" or "gene set enrichment"

We will use the **gage** package and the **pathviewer** package to do this analysis (but there are loads of others).

```{r, message= FALSE}
library(pathview)
library(gage)
library(gageData)
```

Let's see what is in gageData, specifically KEGG pathways:
```{r}
data(kegg.sets.hs)
head(kegg.sets.hs,2)
```

To run our pathway analysis we will use the **gage()** function. It wants two main inputs: a vector of importance (in our case the log2 fold change values); and the gene sets to cehck overlap for. 

```{r}
foldchanges <- res$log2FoldChange
names(foldchanges) <- res$symbol
head(foldchanges)
```

KEGG speaks entrez (i.e. uses ENTREZID format) not gene symbol format. 

```{r}
names(foldchanges) <- res$entrez
head(foldchanges)
```

```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
head(keggres$less, 5)
```

Let's make a figure of one of these pathways with our DEGs highlighted: 

```{r}
pathview(foldchanges, pathway.id = "hsa05310")
```

![](hsa05310.pathview.png)

> Q. Generate and insert a pathway figure for "Graft-versus-host diases" and "Type I diabetes"?

```{r}
pathview(foldchanges, pathway.id = "hsa05332")
pathview(foldchanges, pathway.id = "hsa04940")
```

![](hsa05332.pathview.png)

![](hsa04940.pathview.png)

